[Python] Exception when txt="" in publish_continue_session()

**Version**
0.4.1

**Description**
An exception occurs when providing an empty string to the tts parameters in "publish_continue_session"

**Logs**

```
2019-04-29 11:57:26,862 - INFO : SnipsPlatform.terminate - Snips Platform terminated
FAILED

============================================================================================================== FAILURES ===============================================================================================================
__________________________________________________________________________________________ TestContinueSession.test_publish_continue_session __________________________________________________________________________________________

self = <test_dialogue.TestContinueSession object at 0x10fea47b8>

    def test_publish_continue_session(self):
        """
        Given:
            - An assistant containing the weather application is installed.
            - The user utter "Hey Snips! What is the weather in Tokyo"
            - The assistant catch the intent and continue the session and utter some feedback.
    
        When:
            - The user utter "What is the weather in San Francisco?"
    
        Then:
            - The SearchWeatherForecast intent MUST be published.
            - The forecast_locality slot MUST contain the value San Francisco
            - The session is closed after publishing an end session using Hermes.
        """
    
        # Given
        with WeatherDemoAssistant() as weather_assistant:
    
            params = {
                "hijack": True,
                "no_mike": True,
                "sound_feedback_disabled_default": True,
                "assistant": weather_assistant.assistant_path
            }
    
            with SnipsPlatform(**params) as snips:
    
                intent_name = "searchWeatherForecast"
    
                with IntentListener(intent_name) as intent_listener:
                    snips.audio_server.inject_audio(get_asset("hey_snips-whats_the_weather_in_tokyo.wav"))
    
                    assert intent_listener.exists()
    
                    client: Hermes = intent_listener.queue[0].client
                    msg: IntentMessage = intent_listener.queue[0].message
    
>                   client.publish_continue_session(msg.session_id, "", [intent_name])

client     = <hermes_python.hermes.Hermes object at 0x10ff26320>
intent_listener = <snipsbox.hermes.IntentListener object at 0x10ff262b0>
intent_name = 'searchWeatherForecast'
msg        = <hermes_python.ontology.dialogue.intent.IntentMessage object at 0x10ff26978>
params     = {'assistant': '/var/folders/1g/0sv9v9pn5mb3q1n6szct90t00000gn/T/proj_WEATHER-DEMO-EN/assistant',
 'hijack': True,
 'no_mike': True,
 'sound_feedback_disabled_default': True}
self       = <test_dialogue.TestContinueSession object at 0x10fea47b8>
snips      = <snipsbox.services.SnipsPlatform object at 0x10fea4588>
weather_assistant = <snipsbox.assistant.WeatherDemoAssistant object at 0x10fb741d0>

tests/platform/test_dialogue.py:178: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.virtualenvs/snips-tests/lib/python3.7/site-packages/hermes_python/hermes.py:188: in publish_continue_session
    self.ffi.dialogue.publish_continue_session(continue_session_msg)
../../.virtualenvs/snips-tests/lib/python3.7/site-packages/hermes_python/api/ffi/dialogue.py:135: in publish_continue_session
    message
../../.virtualenvs/snips-tests/lib/python3.7/site-packages/hermes_python/api/ffi/dialogue.py:192: in _call_foreign_function
    ptr_to_foreign_function_argument
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<hermes_python.api.ffi.dialogue.LP_CDialogueFacade object at 0x10e397bf8>, <cparam 'P' (0x10ff1c510)>), kwargs = {}, return_code = 1, empty_string = b''
error_p = <hermes_python.ffi.ontology.LP_c_char_p object at 0x10fec4598>, error_cause = 'could not borrow, unexpected null pointer\n'

    def wrapped_library_call(*args, **kwargs):
        return_code = lib_func(*args, **kwargs)
        if return_code > 0:  # An error occured
            empty_string = "".encode('utf-8')
            error_p = POINTER(c_char_p)(c_char_p(empty_string))
            # Retrieve the last error and put it in the memory location error_p points to
            lib.hermes_get_last_error(error_p)
            error_cause = string_at(error_p.contents).decode('utf-8')
>           raise LibException(error_cause)
E           hermes_python.ffi.LibException: could not borrow, unexpected null pointer

args       = (<hermes_python.api.ffi.dialogue.LP_CDialogueFacade object at 0x10e397bf8>,
 <cparam 'P' (0x10ff1c510)>)
empty_string = b''
error_cause = 'could not borrow, unexpected null pointer\n'
error_p    = <hermes_python.ffi.ontology.LP_c_char_p object at 0x10fec4598>
kwargs     = {}
lib_func   = <_FuncPtr object at 0x10fb54048>
return_code = 1
```

--- a/platforms/hermes-python/documentation/source/HISTORY.rst
+++ b/platforms/hermes-python/documentation/source/HISTORY.rst
@@ -1,5 +1,12 @@
 History
 ==========
+0.5.2 (2019-05-07)
+^^^^^^^^^^^^^^^^^^
+* Fixes nullable fields in Dialogue ontology and brings more type annotations
+
+0.5.1 (2019-05-06)
+^^^^^^^^^^^^^^^^^^
+* introduces new (cli) API to build python wheel that include pre-compiled hermes-mqtt-ffi extension.
 
 0.5.0 (2019-04-19)
 ^^^^^^^^^^^^^^^^^^^

--- a/platforms/hermes-python/hermes_python/__version__.py
+++ b/platforms/hermes-python/hermes_python/__version__.py
@@ -1,5 +1,5 @@
 __title__ = 'hermes_python'
-__version__ = '0.5.1'
+__version__ = '0.5.2'
 __description__ = 'Python bindings for Snips Hermes Protocol'
 __url__ = {
     'Documentation': 'https://hermespython.readthedocs.io/en/latest/',

--- a/platforms/hermes-python/hermes_python/api/ffi/__init__.py
+++ b/platforms/hermes-python/hermes_python/api/ffi/__init__.py
@@ -44,6 +44,7 @@ class FFI(object):
         self.tts.release_facade()
 
     def release_connection(self):
+        utils.hermes_destroy_mqtt_protocol_handler(self._protocol_handler)
         self._protocol_handler = POINTER(CProtocolHandler)()
         self.release_facades()
 

--- a/platforms/hermes-python/hermes_python/api/ffi/__init__.py
+++ b/platforms/hermes-python/hermes_python/api/ffi/__init__.py
@@ -1,6 +1,6 @@
 from ctypes import POINTER, c_char_p, byref
 from ...ffi.ontology import CProtocolHandler, CMqttOptions
-from ...ffi.utils import hermes_protocol_handler_new_mqtt_with_options
+from ...ffi.utils import hermes_protocol_handler_new_mqtt_with_options, hermes_destroy_mqtt_protocol_handler
 from ...ffi import utils, lib
 
 from .dialogue import DialogueFFI
@@ -44,7 +44,7 @@ class FFI(object):
         self.tts.release_facade()
 
     def release_connection(self):
-        utils.hermes_destroy_mqtt_protocol_handler(self._protocol_handler)
+        hermes_destroy_mqtt_protocol_handler(self._protocol_handler)
         self._protocol_handler = POINTER(CProtocolHandler)()
         self.release_facades()
 

--- a/platforms/hermes-python/hermes_python/ffi/ontology/dialogue.py
+++ b/platforms/hermes-python/hermes_python/ffi/ontology/dialogue.py
@@ -26,7 +26,7 @@ class CContinueSessionMessage(Structure):
     @classmethod
     def build(cls, session_id, text, intent_filter, custom_data, slot=None, send_intent_not_recognized=False):
         session_id = session_id.encode('utf-8')
-        text = text.encode('utf-8') if text else None
+        text = text.encode('utf-8')
         intent_filter = [intent_filter_item.encode('utf-8') for intent_filter_item in intent_filter]
 
         c_intent_filter = CStringArray()

--- a/platforms/hermes-python/hermes_python/ffi/utils.py
+++ b/platforms/hermes-python/hermes_python/ffi/utils.py
@@ -11,6 +11,7 @@ from ..ontology import MqttOptions
 
 hermes_protocol_handler_new_mqtt = wrap_library_call(lib.hermes_protocol_handler_new_mqtt)
 hermes_protocol_handler_new_mqtt_with_options = wrap_library_call(lib.hermes_protocol_handler_new_mqtt_with_options)
+hermes_destroy_mqtt_protocol_handler = wrap_library_call(lib.hermes_destroy_mqtt_protocol_handler)
 
 # dialogue
 hermes_protocol_handler_dialogue_facade = wrap_library_call(lib.hermes_protocol_handler_dialogue_facade)

--- a/platforms/hermes-python/hermes_python/ontology/dialogue/session.py
+++ b/platforms/hermes-python/hermes_python/ontology/dialogue/session.py
@@ -125,7 +125,8 @@ class SessionStartedMessage(object):
         :param session_id: Session identifier that was started.
         :param custom_data: Custom data provided in the start session request on.
         :param site_id:  Site where the user interaction is taking place.
-        :param reactivated_from_session_id: This optional field indicates this session is a reactivation of a previously
+        :param reactivated_from_session_id: NOT IMPLEMENTED YET. This feature is coming soon.
+        This optional field indicates this session is a reactivation of a previously
         ended session. This is for example provided when the user continues talking to the platform without saying the
         hotword again after a session was ended.
         """

--- a/platforms/hermes-python/hermes_python/ontology/dialogue/session.py
+++ b/platforms/hermes-python/hermes_python/ontology/dialogue/session.py
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 from builtins import object
 from six.moves import range
+from typing import List, Optional
 
 from hermes_python.ffi.ontology.dialogue import CStartSessionMessageAction, CStartSessionMessageNotification, \
     CSessionInitAction, CSessionInitNotification, CEndSessionMessage, CContinueSessionMessage
@@ -12,8 +13,9 @@ class SessionInit(object):
 
 class SessionInitAction(SessionInit):
     def __init__(self, text=None, intent_filter=list(), can_be_enqueued=True, send_intent_not_recognized=False):
+        # type:(Optional[str], Optional[List[str]], bool, bool) -> None
         """
-        :param text: A text to say to the user
+        :param text: An optional text to say to the user
         :param intent_filter: An optional list of intent name to restrict the parsing of the user response to
         :param can_be_enqueued: An optional boolean to indicate if the session can be enqueued if it can't be started
         immediately (ie there is another running session on the site). The default value is true
@@ -56,6 +58,7 @@ class SessionInitAction(SessionInit):
 
 class SessionInitNotification(SessionInit):
     def __init__(self, text=""):
+        # type:(str) -> None
         """
         The session doesn't expect a response from the user.
         If the session cannot be started, it will enqueued.
@@ -115,13 +118,16 @@ class StartSessionMessage(object):
 
 class SessionStartedMessage(object):
     def __init__(self, session_id, custom_data, site_id, reactivated_from_session_id):
+        # type: (str, Optional[str], str, Optional[str]) -> None
         """
         A message that the handler receives from the Dialogue Manager when a session is started.
 
         :param session_id: Session identifier that was started.
         :param custom_data: Custom data provided in the start session request on.
         :param site_id:  Site where the user interaction is taking place.
-        :param reactivated_from_session_id: This field is left blank voluntarily.
+        :param reactivated_from_session_id: This optional field indicates this session is a reactivation of a previously
+        ended session. This is for example provided when the user continues talking to the platform without saying the
+        hotword again after a session was ended.
         """
         self.session_id = session_id
         self.custom_data = custom_data
@@ -130,9 +136,9 @@ class SessionStartedMessage(object):
 
     @classmethod
     def from_c_repr(cls, c_repr):
-        session_id = c_repr.session_id.decode('utf-8') if c_repr.session_id else None
+        session_id = c_repr.session_id.decode('utf-8')
         custom_data = c_repr.custom_data.decode('utf-8') if c_repr.custom_data else None
-        site_id = c_repr.site_id.decode('utf-8') if c_repr.site_id else None
+        site_id = c_repr.site_id.decode('utf-8')
         reactivated_from_session_id = c_repr.reactivated_from_session_id.decode('utf-8') if c_repr.reactivated_from_session_id else None
         return cls(session_id, custom_data, site_id, reactivated_from_session_id)
 
@@ -142,7 +148,7 @@ class SessionStartedMessage(object):
 
 class EndSessionMessage(object):
     def __init__(self, session_id, text=None):
-        # type: (str, str) -> None
+        # type: (str, Optional[str]) -> None
         """
         :param session_id: The id of the session to end
         :param text: An optional text to say to the user before ending the session
@@ -155,7 +161,7 @@ class EndSessionMessage(object):
 
     @classmethod
     def from_c_repr(cls, c_repr):
-        session_id = c_repr.session_id.decode('utf-8') if c_repr.session_id else None
+        session_id = c_repr.session_id.decode('utf-8')
         text = c_repr.text.decode('utf-8') if c_repr.text else None
         return cls(session_id, text)
 
@@ -165,6 +171,7 @@ class EndSessionMessage(object):
 
 class SessionEndedMessage(object):
     def __init__(self, session_id, custom_data, site_id, termination):
+        # type: (str, Optional[str], str, SessionTermination) -> None
         """
         A message that the handler receives from the Dialogue Manager when a session is ended.
 
@@ -192,6 +199,7 @@ class SessionEndedMessage(object):
 
 class SessionQueuedMessage(object):
     def __init__(self, session_id, custom_data, site_id):
+        # type: (str, Optional[str], str) -> None
         """
         A message that the handler receives from the Dialogue Manager when a session is queued.
 
@@ -236,6 +244,7 @@ class SessionTermination(object):
 
 class ContinueSessionMessage(object):
     def __init__(self, session_id, text, intent_filter, custom_data, send_intent_not_recognized, slot=None):
+        # type: (str, str, List[str], Optional[str], bool, Optional[str]) -> None
         """
         :param session_id: Identifier of the dialogue session during which this intent was parsed.
         :param text:
@@ -283,6 +292,7 @@ class ContinueSessionMessage(object):
 
 class IntentNotRecognizedMessage(object):
     def __init__(self, site_id, session_id, input, custom_data, confidence_score):
+        # type: (str, str, Optional[str], Optional[str], float) -> None
         """
         A message that the handler receives from the Dialogue manager when an intent is not recognized and that the
         session was initialized with the intent_not_recognized flag turned on.

--- a/platforms/hermes-python/tests/api/ffi/test_ffi.py
+++ b/platforms/hermes-python/tests/api/ffi/test_ffi.py
@@ -42,8 +42,8 @@ def test_establish_connection_calls_api_subsets(hermes_protocol_handler_new_mqtt
     ffi.injection.initialize_facade.assert_called_once()
     ffi.tts.initialize_facade.assert_called_once()
 
-
-def test_release_connection_calls_api_subsets():
+@mock.patch("hermes_python.api.ffi.hermes_destroy_mqtt_protocol_handler")
+def test_release_connection_calls_api_subsets(hermes_destroy_mqtt_protocol_handler):
     ffi = FFI()
 
     # Here, you have to mock every API subset of Hermes Protocol
@@ -57,8 +57,11 @@ def test_release_connection_calls_api_subsets():
     ffi.dialogue.release_facade.assert_called_once()
     ffi.sound_feedback.release_facade.assert_called_once()
 
+    hermes_destroy_mqtt_protocol_handler.assert_called_once()
+
 
 @mock.patch("hermes_python.api.ffi.hermes_protocol_handler_new_mqtt_with_options")
+@mock.patch("hermes_python.api.ffi.hermes_destroy_mqtt_protocol_handler")
 @mock.patch("hermes_python.api.ffi.DialogueFFI")
 @mock.patch("hermes_python.api.ffi.SoundFeedBackFFI")
 @mock.patch("hermes_python.api.ffi.InjectionFFI")
@@ -69,6 +72,7 @@ class ConnectionTest(object):
                                                 injectionFFI,
                                                 soundfeedbackFFI,
                                                 dialogueFFI,
+                                                hermes_destroy_mqtt_protocol_handler,
                                                 hermes_protocol_handler_new_mqtt):
         ffi = FFI()
         mqtt_opts = MqttOptions()
@@ -77,12 +81,14 @@ class ConnectionTest(object):
         hermes_protocol_handler_new_mqtt.assert_called_once()
         ffi.dialogue.initialize_facade.assert_called_once()
         ffi.sound_feedback.initialize_facade.assert_called_once()
+        hermes_destroy_mqtt_protocol_handler.assert_called_once()
 
     def test_release_connection_sucessful(self,
                                           ttsFFI,
                                           injectionFFI,
                                           soundfeedbackFFI,
                                           dialogueFFI,
+                                          hermes_destroy_mqtt_protocol_handler,
                                           hermes_protocol_handler_new_mqtt):
         ffi = FFI()
         mqtt_opts = MqttOptions()
@@ -96,6 +102,8 @@ class ConnectionTest(object):
         ffi.dialogue.release_facade.assert_called_once()
         ffi.sound_feedback.initialize_facade.assert_called_once()
         ffi.sound_feedback.release_facade.assert_called_once()
+
+        hermes_destroy_mqtt_protocol_handler.assert_called_once()
 
 
 

--- a/platforms/hermes-python/tests/test_hermes.py
+++ b/platforms/hermes-python/tests/test_hermes.py
@@ -40,8 +40,10 @@ def test_context_manager_enter_calls_ffi_api():
 @mock.patch("hermes_python.api.ffi.feedback.hermes_protocol_handler_sound_feedback_facade")
 @mock.patch("hermes_python.api.ffi.dialogue.hermes_drop_dialogue_facade")
 @mock.patch("hermes_python.api.ffi.dialogue.hermes_protocol_handler_dialogue_facade")
+@mock.patch("hermes_python.api.ffi.hermes_destroy_mqtt_protocol_handler")
 @mock.patch("hermes_python.api.ffi.hermes_protocol_handler_new_mqtt_with_options")
 def test_context_manager_enter_exit(hermes_protocol_handler_new_mqtt,
+                                    hermes_destroy_mqtt_protocol_handler,
                                     hermes_protocol_handler_dialogue_facade, hermes_drop_dialogue_facade,
                                     hermes_protocol_handler_sound_feedback_facade, hermes_drop_sound_feedback_facade,
                                     hermes_protocol_handler_injection_facade, hermes_drop_injection_facade,
@@ -62,6 +64,8 @@ def test_context_manager_enter_exit(hermes_protocol_handler_new_mqtt,
 
     hermes_protocol_handler_tts_facade.assert_called_once()
     hermes_drop_tts_facade.assert_called_once()
+
+    hermes_destroy_mqtt_protocol_handler.assert_called_once()
 
 
 @mock.patch("hermes_python.api.ffi.feedback.hermes_protocol_handler_sound_feedback_facade")

